service_handler.go

```diff
@@ -0,0 +1,423 @@
+/*
+Package cynic monitors you from the ceiling.
+
+Copyright 2018 Simon Symeonidis (psyomn)
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package cynic
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"sync"
+	"time"
+)
+
+const (
+	// StopService is the signal to stop the running querying service
+	StopService = iota
+
+	// AddService adds a service to a running cynic instance
+	AddService = iota
+
+	// DeleteService removes a service from a running cynic instance
+	DeleteService = iota
+)
+
+// AlertFunc defines the hook signature for alert messages
+type AlertFunc = func([]AlertMessage)
+
+// AlertMessage defines a simple alert structure that can be used by
+// users of the library, and decide how to show information about the
+// alerts.
+type AlertMessage struct {
+	Response      interface{} `json:"response_text"`
+	Endpoint      string      `json:"endpoint"`
+	Now           string      `json:"now"`
+	CynicHostname string      `json:"cynic_hostname"`
+}
+
+// Session is the configuration a cynic instance requires to start
+// running and working
+type Session struct {
+	StatusPort string
+	SlackHook  *string
+	Services   []Service
+	Alerter    AlertFunc
+	AlertTime  int
+}
+
+// Service is some http service location that should be queried in a
+// specified amount of time.
+//
+// Hooks: it is possible to assign a hook per service. When the
+// service successfully fires and finishes, the retrieved information
+// will be passed to the hook. The hook can do whatever, and should
+// return a encodable JSON structure.
+//
+// The hook can do whatever with said information. User defined things
+// happen in the hook, and the hook returns a structure ready to be
+// encoded into a JSON object, inserted in the sync map, and then
+// served back to the client whenever queried.
+//
+// - A service is an HTTP endpoint
+// - A service can have many:
+//   - hooks (that can act as contracts)
+type Service struct {
+	URL        url.URL
+	Secs       int
+	hooks      []interface{}
+	ticker     *time.Ticker
+	running    bool
+	tickerChan chan int
+	immediate  bool
+	offset     int
+}
+
+// TODO make sure that this is used everywhere.
+type serviceError struct {
+	Error string `json:"error"`
+}
+
+// AddressBook contains all the required services inside a map.
+type AddressBook struct {
+	entries      map[string]*Service
+	statusServer StatusServer
+	Mutex        *sync.Mutex
```

Comment:
I still have to dig a little as to why, but I would heartily recommend that your mutex not be visible or used outside of the class that uses it. It's much easier to manage the complexity of mutexes if you reduce their possible usage scope as much as possible.

Reply:
Yep, this was supposed to be private, haven't gone around it yet.